keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,980 @@
+way, the cluster can now grow by an additional 16 nodes.
+
+#### Add nodes with higher or lower capabilities
+
+A user created a cluster with an ample sized cluster CIDR. All the initial nodes
+are of uniform capacity capable of running a maximum of 256 pods and they are
+each assigned a /24 pod CIDR. The user is planning to add more nodes to the
+system which are capable of running 500 pods. However, they cannot take
+advantage of the additional capacity because all nodes are assigned a /24 pod
+CIDR. With this enhancement the user configures a new allocation which uses the
+original cluster CIDR but allocates a /23 instead of a /24 to each node. They
+use the node selector to allocate these IPs only to the nodes with the higher
+capacity.
+
+#### Provision discontiguous ranges
+
+A user wants to create a cluster with 32 nodes each with a capacity to run 256
+pods. This means that each node needs a /24 pod CIDR range and they need a total
+range of /19. However, there aren't enough contiguous IPs in the user's network.
+They can find 4 free ranges of size /21 but no single contiguous /19 range.
+
+Using this enhancement, the user creates 4 different CIDR configurations each
+with a /21 range. The CIDR allocator allocates a /24 range from any of these /21
+ranges to the nodes and the user can now create the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+This feature does not expand the ability of the NodeIPAM controller to change
+the `Node.Spec.PodCIDRs` field. Once that field is set, either by the controller
+or a third party, it will be treated as immutable. This is particularly relevant
+in situtaitons where users start modifying or deleting the `ClusterCidrConfig`.
+Under no circumstances will the controller attempt to revoke the allocated
+CIDRs (more details on this are discussed below).
+
+### Risks and Mitigations
+
+-   Racing kube-controller-managers. If multiples of the controller are running
+    (as in a HA control plane), how do they coordinate?
+    -   The controllers will coordinate using the existing
+        kube-controller-manager leader election.
+
+## Design Details
+
+### New Resource
+
+```go
+type ClusterCIDRConfig struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+
+  Spec ClusterCIDRConfigSpec
+  Status ClusterCIDRConfigStatus
+}
+
+type ClusterCIDRConfigSpec {
+    // An IP block in CIDR notation ("10.0.0.0/8", "fd12:3456:789a:1::/64")
+    // +required
+    CIDR string
+
+    // This defines which nodes the config is applicable to. A nil selector can
+    // be applied to any node.
+    // +optional
+    NodeSelector *v1.LabelSelector
+
+    // Netmask size (e.g. 25 -> "/25") to allocate to a node.
+    // Users would have to ensure that the kubelet doesn't try to schedule
+    // more pods than are supported by the node's netmask (i.e. the kubelet's
+    // --max-pods flag)
+    // +required
+    PerNodeMaskSize int
+
+    // Is this a good idea?
```

Comment:
What is the intent of this?

Reply:
Whoops, that's me playing around. Didn't mean to upload that.